feat: Hero NFT redesign confirmation

digiwand opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

- Abstracted Hero presentational component
- Create HeroNFT component 
- (TokenHero to be renamed and relocated to a similar path: hero/token in a follow-up PR)

No QA Needed label added as this will be QA'd as a whole with the rest of the transfer redesign

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/5020

## **Manual testing steps**

1. Turn on transaction redesign feature flags
2. Go to NFT panel
3. Send NFT
4. Observe top hero component

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

![CleanShot 2025-05-27 at 23 24 30](https://github.com/user-attachments/assets/2962285f-96d6-4386-903d-f8c0ae5b42d1)
![CleanShot 2025-05-27 at 19 23 41](https://github.com/user-attachments/assets/0de52ce3-58c3-457e-b8c8-e2df74979342)

https://github.com/user-attachments/assets/d200ee20-73c6-42b6-a224-18e778ff689b

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

digiwand committed:
refactor: abstract Hero from TokenHero

digiwand committed:
feat: Hero title increase padding by 4

digiwand committed:
refactor: rm the defaulted BodyMd Text variant

digiwand committed:
feat: support HeroNft transfer redesign

digiwand committed:
style: mv nft hooks

digiwand committed:
test: create useIsNft test

digiwand committed:
test: create confirmation Hero test

digiwand committed:
test: create mock state data nft

digiwand committed:
test: create HeroNft test (WIP)

digiwand committed:
test: create HeroNft test (WIP)

digiwand committed:
feat: add onPress to HeroNft image

digiwand committed:
test: add useNFT test

digiwand committed:
refactor: HeroNft rm isFirstPartyContractName logic

digiwand committed:
refactor: replace useDisplayNames hook in useNft

digiwand committed:
test: add HeroNft tests and cleanup

digiwand committed:
fix: rm lint unused import + format selection

OGPoyraz commented on app/components/Views/confirmations/components/hero/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { Hero } from './hero';
+// export { HeroNft } from './nft/nft';
```
```suggestion
```

OGPoyraz commented on app/components/Views/confirmations/hooks/nft/useNft.ts:
```diff
@@ -0,0 +1,49 @@
+
+export interface UseNftResponse {
+  chainId: string;
+
```
```suggestion
```

OGPoyraz commented on app/components/Views/confirmations/components/info/transfer/transfer.tsx:
```diff
@@ -26,9 +28,15 @@ const Transfer = () => {
     <View>
-      <TokenHero />
+      {isNft ? (
+        <HeroNft />
+      ) : (
+        <TokenHero />
```
Minor, shall we do a minor rename here? Either `TokenHero` to `HeroToken` or `HeroNft` to `NftHero`?

digiwand replied:
yep! the plan is to rename and relocate TokenHero to the new file path components/hero. I wanted to refrain from this update now to ensure the code changes are easier to review.

thanks for bringing this up here.

OGPoyraz commented on app/components/Views/confirmations/hooks/nft/useNft.ts:
```diff
@@ -0,0 +1,49 @@
+
+  const nftContract = useNftContract(chainId, tokenAddress);
+
+  return {
+    chainId: toHex(chainId).toString(),
+    name: nftContract?.name,
+    nft,
+    tokenId,
+  };
```
This NFT hooks look brilliant 🤩

digiwand replied:
ty! found this interesting

`collectiblesSelector` and `NftController.findNftByAddressAndTokenId` (not currently used in metamask-mobile), both returned an NFT object with metadata. `collectiblesSelector` reads from `nftControllerState.allNfts`. The name they provide appends its ID, e.g., "Example Nft #777".

we needed the name of the NFT without the appended number. This can be determined by retrieving the NFT contract details.

`useDisplayName` aggregates multiple assets to resolve the name. This returned the name we desired without the appended ID, "Example Nft". Since it is pulling from the contract, it does not return the image we need for ERC721. 

In our case, the hook `useDisplayName` derived the nft contract from `useWatchedNFTNames`. `useWatchedNFTNames` pulled from `selectAllNftContracts`, which reads from `nftControllerState.allNftContracts`.

I replaced the `useDisplayName` and `useWatchedNFTNames` logic with the creation of useNftContract to directly find the contract by address and chain id.

OGPoyraz replied:
That totally make sense

OGPoyraz commented on app/components/Views/confirmations/components/hero/nft/nft.tsx:
As discussed before let's use a flat structure here.

OGPoyraz commented on app/components/Views/confirmations/components/rows/transactions/token-hero/token-hero.tsx:
```diff
@@ -1,119 +1,75 @@
-          />
-        )}
-      </View>
+      <Hero
+        componentAsset={<AvatarTokenWithNetworkBadge />}
+        title={<AssetAmount amount={amount} styles={styles} />}
+        subtitle={showFiatOnTestnets ? fiat : undefined}
+        tooltipModalProps={{
+          content: amountPrecise,
+          isEnabled: isRoundedAmount,
+          hasTooltip: true,
+          testId: 'token-hero-amount',
+          title: strings('send.amount'),
+        }}
+      />
```
Can we not create another component here?

Then it would be much cleaner if we would just have two components as you put in the render method of `token-hero`, either `NFTHero` or `TokenHero` as you already done it in `transfer.tsx` and no more file hop in those components?

digiwand replied:
created components/hero/transfer

OGPoyraz reviewed (changes requested):
This PR looks awesome, nothing major on code changes! 

Just two small re-organisation request and we are good to go with it

OGPoyraz commented on app/components/Views/confirmations/components/info/transfer/transfer.tsx:
```diff
@@ -26,9 +28,15 @@ const Transfer = () => {
   return (
     <View>
-      <TokenHero />
+      {isNft ? (
```
Ideally let's have a single `components/rows/transactions/hero-row` component that either renders `TokenHero` or `NftHero`, so we don't have to put render logic directly here. There some benefits to have row single row like this - we can easily check

OGPoyraz replied:
It's totally understandable that confusion on the sections vs rows. But in file structure we decided to use `row` instead of `section`.

You may have a question of "when we define the second level folder (when to use `components/FOLDER`)" . That's totally tied to domain folders. For example `gas` or `7702`. In that case we don't want to have `hero` folder containing multiple components.

To be honest - I fall into same pattern while doing gas components and decided to use real flattened components. 
So I mean more like separated `hero-base` (assuming this is for ERC20 and native) , `hero-nft` and `row/hero-row`(as we use it as row)

digiwand replied:
sounds good to me! ty @OGPoyraz

digiwand committed:
comment: rm unused export HeroNft

Co-authored-by: OGPoyraz <omergoktugpoyraz@gmail.com>

digiwand committed:
mv: hero -> header/hero

digiwand committed:
refactor: rename header/hero -> header/hero/base

digiwand committed:
refactor: rename Hero -> HeroBase

digiwand committed:
refactor: create HeroTransfer component

digiwand committed:
refactor: mv header/hero -> hero

digiwand committed:
refactor: mv hero/base -> hero

digiwand committed:
refactor: mv hero/nft -> hero-nft

digiwand committed:
refactor: mv hero -> UI/hero

digiwand committed:
Merge branch 'main' into feat-confirm-hero-nft-2

digiwand committed:
fix: lint paths

OGPoyraz committed:
Reorg hero components (#16057)

Reorg of hero components 

![Screenshot 2025-06-04 at 09 53
45](https://github.com/user-attachments/assets/6bcf0a0e-bcc2-4766-af95-8b0d8cfb47a5)

OGPoyraz committed:
Merge branch 'main' into feat-confirm-hero-nft-2

digiwand commented on app/components/Views/confirmations/components/UI/hero/hero.tsx:
```diff
@@ -0,0 +1,108 @@
+
+  return (
+    <View style={styles.base}>
+      {componentAsset}
```
we could consider changing this to {children}, or renaming componentAsset. It's so far used to display the asset, but we could be make this more generic to support other cases

OGPoyraz replied:
Make sense @digiwand , can you do that change so you can have the latest commit - so I can approve and we merge this PR?
